core.ts

```diff
@@ -0,0 +1,1039 @@
+	if (obj == null) {
+		return false
+	}
+	for (const method of requiredUtilMethods) {
+		if (obj[method] == null) {
+			return false
+		}
+	}
+	return true
+}
+const isPromiseRejector = (obj: any): obj is PinejsClientCoreFactory.PromiseRejector => {
+	return obj != null && obj.reject != null
+}
+const isValidOption = (key: string): key is keyof PinejsClientCoreFactory.ODataOptions => {
+	return key === '$filter' ||
+		key === '$expand' ||
+		key === '$orderby' ||
+		key === '$top' ||
+		key === '$skip' ||
+		key === '$select'
+}
+
+// Workaround the fact that `setInterval` returns a different type in nodejs vs browsers
+// TODO: typescript 2.8 will introduce `ReturnType` as a better way to do this
+const _setIntervalResult = false as true && setInterval(() => {}, 0)
+
+type PollOnObj = {
+	unsubscribe: () => void
+}
+class Poll<PromiseResult extends PromiseLike<number | PinejsClientCoreFactory.AnyObject | PinejsClientCoreFactory.AnyObject[]> = Promise<number | PinejsClientCoreFactory.AnyObject | PinejsClientCoreFactory.AnyObject[]>> {
+	private subscribers: {
+		error: Array<(response: PromiseResult) => void>,
+		data: Array<(err: any) => void>,
+	} = {
+		error: [],
+		data: [],
+	}
+
+	private stopped = false
+	private pollInterval?: typeof _setIntervalResult
+
+	private requestFn: null | (() => PromiseResult)
+
+	constructor(
+		requestFn: () => PromiseResult,
+		private intervalTime = 10000
+	) {
+		this.requestFn = requestFn
+		this.start()
+	}
+
+	setPollInterval(intervalTime: number) {
+		this.intervalTime = intervalTime
+		this.stop()
+		this.start()
+	}
+
+	runRequest() {
+		if (this.stopped || this.requestFn == null) {
+			return
+		}
+		this.requestFn()
+		.then((response) => {
+			if (this.stopped) {
+				return
+			}
+
+			// Catch errors in event subscribers so that they don't trigger
+			// the 'catch' below, and that subsequent subscribers will still
+			// be called
+			this.subscribers.data.forEach((fn) => {
+				try {
+					fn(response)
+				} catch (error) {
+					console.error('pinejs-client error: Caught error in data event subscription:', error)
+				}
+			})
+
+			return null
+		}, (err: any) => {
+			if (this.stopped) {
+				return
+			}
+
+			this.subscribers.error.forEach((fn) => {
+				try {
+					fn(err)
+				} catch (error) {
+					console.error('pinejs-client error: Caught error in error event subscription:', error)
+				}
+			})
+
+			return null
+		})
+	}
+
+	on(name: 'error', fn: (response: PromiseResult) => void): PollOnObj
+	on(name: 'data', fn: (err: any) => void): PollOnObj
+	on(name: keyof Poll['subscribers'], fn: (value: any) => void): PollOnObj {
+		const subscribers = this.subscribers[name] as Array<(value: any) => void>
+		const index = subscribers.push(fn) - 1
+
+		return {
+			unsubscribe: () => delete this.subscribers[name][index]
+		}
+	}
+
+	start() {
+		this.stopped = false
+		this.runRequest()
+		this.pollInterval = setInterval(
+			() => this.runRequest(),
+			this.intervalTime
+		)
```

Comment:
This seems slightly naive and can be problematic if the request takes longer than `this.intervalTime` to complete, isn't it so? Maybe `runRequest` calling itself after it finishes is better?

Reply:
Yeah I think that would be preferable, I just left it the way it was when converting though - @LucianBuzzo does it cause you any problems to have it switch to a `setTimeout` after the request finishes instead?  That way it would mean we don't stack up more requests whilst the previous ones are still in-process

Reply:
@Page- Yeah for sure, that sounds like a good idea

Reply:
Done